// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: bulbasaur/api/bulbasaur.proto

package bulbasaur

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on TokenInfo with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TokenInfo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TokenInfo with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in TokenInfoMultiError, or nil
// if none found.
func (m *TokenInfo) ValidateAll() error {
	return m.validate(true)
}

func (m *TokenInfo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for SafeId

	// no validation rules for RefreshToken

	// no validation rules for AccessToken

	// no validation rules for Role

	// no validation rules for UserId

	if len(errors) > 0 {
		return TokenInfoMultiError(errors)
	}

	return nil
}

// TokenInfoMultiError is an error wrapping multiple validation errors returned
// by TokenInfo.ValidateAll() if the designated constraints aren't met.
type TokenInfoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TokenInfoMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TokenInfoMultiError) AllErrors() []error { return m }

// TokenInfoValidationError is the validation error returned by
// TokenInfo.Validate if the designated constraints aren't met.
type TokenInfoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TokenInfoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TokenInfoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TokenInfoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TokenInfoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TokenInfoValidationError) ErrorName() string { return "TokenInfoValidationError" }

// Error satisfies the builtin error interface
func (e TokenInfoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTokenInfo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TokenInfoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TokenInfoValidationError{}

// Validate checks the field values on User with the rules defined in the proto
// definition for this message. If any rules are violated, the first error
// encountered is returned, or nil if there are no violations.
func (m *User) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on User with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in UserMultiError, or nil if none found.
func (m *User) ValidateAll() error {
	return m.validate(true)
}

func (m *User) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Username

	// no validation rules for Email

	// no validation rules for Role

	// no validation rules for Id

	if m.Metadata != nil {

		if all {
			switch v := interface{}(m.GetMetadata()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, UserValidationError{
						field:  "Metadata",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, UserValidationError{
						field:  "Metadata",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetMetadata()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return UserValidationError{
					field:  "Metadata",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return UserMultiError(errors)
	}

	return nil
}

// UserMultiError is an error wrapping multiple validation errors returned by
// User.ValidateAll() if the designated constraints aren't met.
type UserMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UserMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UserMultiError) AllErrors() []error { return m }

// UserValidationError is the validation error returned by User.Validate if the
// designated constraints aren't met.
type UserValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UserValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UserValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UserValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UserValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UserValidationError) ErrorName() string { return "UserValidationError" }

// Error satisfies the builtin error interface
func (e UserValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUser.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UserValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UserValidationError{}

// Validate checks the field values on MeResponse with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *MeResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on MeResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in MeResponseMultiError, or
// nil if none found.
func (m *MeResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *MeResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetUser()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, MeResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, MeResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return MeResponseValidationError{
				field:  "User",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return MeResponseMultiError(errors)
	}

	return nil
}

// MeResponseMultiError is an error wrapping multiple validation errors
// returned by MeResponse.ValidateAll() if the designated constraints aren't met.
type MeResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m MeResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m MeResponseMultiError) AllErrors() []error { return m }

// MeResponseValidationError is the validation error returned by
// MeResponse.Validate if the designated constraints aren't met.
type MeResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e MeResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e MeResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e MeResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e MeResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e MeResponseValidationError) ErrorName() string { return "MeResponseValidationError" }

// Error satisfies the builtin error interface
func (e MeResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sMeResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = MeResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = MeResponseValidationError{}

// Validate checks the field values on ListUsersRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListUsersRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListUsersRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListUsersRequestMultiError, or nil if none found.
func (m *ListUsersRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListUsersRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListUsersRequestMultiError(errors)
	}

	return nil
}

// ListUsersRequestMultiError is an error wrapping multiple validation errors
// returned by ListUsersRequest.ValidateAll() if the designated constraints
// aren't met.
type ListUsersRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListUsersRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListUsersRequestMultiError) AllErrors() []error { return m }

// ListUsersRequestValidationError is the validation error returned by
// ListUsersRequest.Validate if the designated constraints aren't met.
type ListUsersRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListUsersRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListUsersRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListUsersRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListUsersRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListUsersRequestValidationError) ErrorName() string { return "ListUsersRequestValidationError" }

// Error satisfies the builtin error interface
func (e ListUsersRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListUsersRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListUsersRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListUsersRequestValidationError{}

// Validate checks the field values on ListUsersResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListUsersResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListUsersResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListUsersResponseMultiError, or nil if none found.
func (m *ListUsersResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListUsersResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetUsers() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListUsersResponseValidationError{
						field:  fmt.Sprintf("Users[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListUsersResponseValidationError{
						field:  fmt.Sprintf("Users[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListUsersResponseValidationError{
					field:  fmt.Sprintf("Users[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListUsersResponseMultiError(errors)
	}

	return nil
}

// ListUsersResponseMultiError is an error wrapping multiple validation errors
// returned by ListUsersResponse.ValidateAll() if the designated constraints
// aren't met.
type ListUsersResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListUsersResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListUsersResponseMultiError) AllErrors() []error { return m }

// ListUsersResponseValidationError is the validation error returned by
// ListUsersResponse.Validate if the designated constraints aren't met.
type ListUsersResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListUsersResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListUsersResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListUsersResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListUsersResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListUsersResponseValidationError) ErrorName() string {
	return "ListUsersResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListUsersResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListUsersResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListUsersResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListUsersResponseValidationError{}

// Validate checks the field values on SignUpRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SignUpRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignUpRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SignUpRequestMultiError, or
// nil if none found.
func (m *SignUpRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SignUpRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Role

	if all {
		switch v := interface{}(m.GetMetadata()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SignUpRequestValidationError{
					field:  "Metadata",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SignUpRequestValidationError{
					field:  "Metadata",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMetadata()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SignUpRequestValidationError{
				field:  "Metadata",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	switch v := m.Credential.(type) {
	case *SignUpRequest_Local_:
		if v == nil {
			err := SignUpRequestValidationError{
				field:  "Credential",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetLocal()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, SignUpRequestValidationError{
						field:  "Local",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, SignUpRequestValidationError{
						field:  "Local",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetLocal()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return SignUpRequestValidationError{
					field:  "Local",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *SignUpRequest_Google_:
		if v == nil {
			err := SignUpRequestValidationError{
				field:  "Credential",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetGoogle()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, SignUpRequestValidationError{
						field:  "Google",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, SignUpRequestValidationError{
						field:  "Google",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetGoogle()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return SignUpRequestValidationError{
					field:  "Google",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if m.DeviceId != nil {
		// no validation rules for DeviceId
	}

	if m.DeviceName != nil {
		// no validation rules for DeviceName
	}

	if len(errors) > 0 {
		return SignUpRequestMultiError(errors)
	}

	return nil
}

// SignUpRequestMultiError is an error wrapping multiple validation errors
// returned by SignUpRequest.ValidateAll() if the designated constraints
// aren't met.
type SignUpRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignUpRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignUpRequestMultiError) AllErrors() []error { return m }

// SignUpRequestValidationError is the validation error returned by
// SignUpRequest.Validate if the designated constraints aren't met.
type SignUpRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignUpRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignUpRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignUpRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignUpRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignUpRequestValidationError) ErrorName() string { return "SignUpRequestValidationError" }

// Error satisfies the builtin error interface
func (e SignUpRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignUpRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignUpRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignUpRequestValidationError{}

// Validate checks the field values on SignUpResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SignUpResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignUpResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SignUpResponseMultiError,
// or nil if none found.
func (m *SignUpResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *SignUpResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetUser()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SignUpResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SignUpResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SignUpResponseValidationError{
				field:  "User",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetTokenInfo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SignUpResponseValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SignUpResponseValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTokenInfo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SignUpResponseValidationError{
				field:  "TokenInfo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return SignUpResponseMultiError(errors)
	}

	return nil
}

// SignUpResponseMultiError is an error wrapping multiple validation errors
// returned by SignUpResponse.ValidateAll() if the designated constraints
// aren't met.
type SignUpResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignUpResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignUpResponseMultiError) AllErrors() []error { return m }

// SignUpResponseValidationError is the validation error returned by
// SignUpResponse.Validate if the designated constraints aren't met.
type SignUpResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignUpResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignUpResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignUpResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignUpResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignUpResponseValidationError) ErrorName() string { return "SignUpResponseValidationError" }

// Error satisfies the builtin error interface
func (e SignUpResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignUpResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignUpResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignUpResponseValidationError{}

// Validate checks the field values on SignInRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SignInRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignInRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SignInRequestMultiError, or
// nil if none found.
func (m *SignInRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SignInRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	switch v := m.Credential.(type) {
	case *SignInRequest_Local_:
		if v == nil {
			err := SignInRequestValidationError{
				field:  "Credential",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetLocal()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, SignInRequestValidationError{
						field:  "Local",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, SignInRequestValidationError{
						field:  "Local",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetLocal()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return SignInRequestValidationError{
					field:  "Local",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *SignInRequest_Google_:
		if v == nil {
			err := SignInRequestValidationError{
				field:  "Credential",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetGoogle()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, SignInRequestValidationError{
						field:  "Google",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, SignInRequestValidationError{
						field:  "Google",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetGoogle()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return SignInRequestValidationError{
					field:  "Google",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if m.DeviceId != nil {
		// no validation rules for DeviceId
	}

	if m.DeviceName != nil {
		// no validation rules for DeviceName
	}

	if len(errors) > 0 {
		return SignInRequestMultiError(errors)
	}

	return nil
}

// SignInRequestMultiError is an error wrapping multiple validation errors
// returned by SignInRequest.ValidateAll() if the designated constraints
// aren't met.
type SignInRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignInRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignInRequestMultiError) AllErrors() []error { return m }

// SignInRequestValidationError is the validation error returned by
// SignInRequest.Validate if the designated constraints aren't met.
type SignInRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignInRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignInRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignInRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignInRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignInRequestValidationError) ErrorName() string { return "SignInRequestValidationError" }

// Error satisfies the builtin error interface
func (e SignInRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignInRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignInRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignInRequestValidationError{}

// Validate checks the field values on SignInResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *SignInResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignInResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in SignInResponseMultiError,
// or nil if none found.
func (m *SignInResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *SignInResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetUser()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SignInResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SignInResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SignInResponseValidationError{
				field:  "User",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetTokenInfo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, SignInResponseValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, SignInResponseValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTokenInfo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return SignInResponseValidationError{
				field:  "TokenInfo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return SignInResponseMultiError(errors)
	}

	return nil
}

// SignInResponseMultiError is an error wrapping multiple validation errors
// returned by SignInResponse.ValidateAll() if the designated constraints
// aren't met.
type SignInResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignInResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignInResponseMultiError) AllErrors() []error { return m }

// SignInResponseValidationError is the validation error returned by
// SignInResponse.Validate if the designated constraints aren't met.
type SignInResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignInResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignInResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignInResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignInResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignInResponseValidationError) ErrorName() string { return "SignInResponseValidationError" }

// Error satisfies the builtin error interface
func (e SignInResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignInResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignInResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignInResponseValidationError{}

// Validate checks the field values on RefreshTokenRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *RefreshTokenRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RefreshTokenRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// RefreshTokenRequestMultiError, or nil if none found.
func (m *RefreshTokenRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *RefreshTokenRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTokenInfo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RefreshTokenRequestValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RefreshTokenRequestValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTokenInfo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RefreshTokenRequestValidationError{
				field:  "TokenInfo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return RefreshTokenRequestMultiError(errors)
	}

	return nil
}

// RefreshTokenRequestMultiError is an error wrapping multiple validation
// errors returned by RefreshTokenRequest.ValidateAll() if the designated
// constraints aren't met.
type RefreshTokenRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RefreshTokenRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RefreshTokenRequestMultiError) AllErrors() []error { return m }

// RefreshTokenRequestValidationError is the validation error returned by
// RefreshTokenRequest.Validate if the designated constraints aren't met.
type RefreshTokenRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RefreshTokenRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RefreshTokenRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RefreshTokenRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RefreshTokenRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RefreshTokenRequestValidationError) ErrorName() string {
	return "RefreshTokenRequestValidationError"
}

// Error satisfies the builtin error interface
func (e RefreshTokenRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRefreshTokenRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RefreshTokenRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RefreshTokenRequestValidationError{}

// Validate checks the field values on RefreshTokenResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *RefreshTokenResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RefreshTokenResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// RefreshTokenResponseMultiError, or nil if none found.
func (m *RefreshTokenResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *RefreshTokenResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTokenInfo()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RefreshTokenResponseValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RefreshTokenResponseValidationError{
					field:  "TokenInfo",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTokenInfo()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RefreshTokenResponseValidationError{
				field:  "TokenInfo",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return RefreshTokenResponseMultiError(errors)
	}

	return nil
}

// RefreshTokenResponseMultiError is an error wrapping multiple validation
// errors returned by RefreshTokenResponse.ValidateAll() if the designated
// constraints aren't met.
type RefreshTokenResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RefreshTokenResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RefreshTokenResponseMultiError) AllErrors() []error { return m }

// RefreshTokenResponseValidationError is the validation error returned by
// RefreshTokenResponse.Validate if the designated constraints aren't met.
type RefreshTokenResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RefreshTokenResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RefreshTokenResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RefreshTokenResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RefreshTokenResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RefreshTokenResponseValidationError) ErrorName() string {
	return "RefreshTokenResponseValidationError"
}

// Error satisfies the builtin error interface
func (e RefreshTokenResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRefreshTokenResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RefreshTokenResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RefreshTokenResponseValidationError{}

// Validate checks the field values on UpdateMetadataRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateMetadataRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateMetadataRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateMetadataRequestMultiError, or nil if none found.
func (m *UpdateMetadataRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateMetadataRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetMetadata()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateMetadataRequestValidationError{
					field:  "Metadata",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateMetadataRequestValidationError{
					field:  "Metadata",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMetadata()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateMetadataRequestValidationError{
				field:  "Metadata",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdateMetadataRequestMultiError(errors)
	}

	return nil
}

// UpdateMetadataRequestMultiError is an error wrapping multiple validation
// errors returned by UpdateMetadataRequest.ValidateAll() if the designated
// constraints aren't met.
type UpdateMetadataRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateMetadataRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateMetadataRequestMultiError) AllErrors() []error { return m }

// UpdateMetadataRequestValidationError is the validation error returned by
// UpdateMetadataRequest.Validate if the designated constraints aren't met.
type UpdateMetadataRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateMetadataRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateMetadataRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateMetadataRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateMetadataRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateMetadataRequestValidationError) ErrorName() string {
	return "UpdateMetadataRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateMetadataRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateMetadataRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateMetadataRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateMetadataRequestValidationError{}

// Validate checks the field values on EmailVerificationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *EmailVerificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on EmailVerificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// EmailVerificationRequestMultiError, or nil if none found.
func (m *EmailVerificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *EmailVerificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Email

	if len(errors) > 0 {
		return EmailVerificationRequestMultiError(errors)
	}

	return nil
}

// EmailVerificationRequestMultiError is an error wrapping multiple validation
// errors returned by EmailVerificationRequest.ValidateAll() if the designated
// constraints aren't met.
type EmailVerificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m EmailVerificationRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m EmailVerificationRequestMultiError) AllErrors() []error { return m }

// EmailVerificationRequestValidationError is the validation error returned by
// EmailVerificationRequest.Validate if the designated constraints aren't met.
type EmailVerificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e EmailVerificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e EmailVerificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e EmailVerificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e EmailVerificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e EmailVerificationRequestValidationError) ErrorName() string {
	return "EmailVerificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e EmailVerificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sEmailVerificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = EmailVerificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = EmailVerificationRequestValidationError{}

// Validate checks the field values on ResetCodeVerificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ResetCodeVerificationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ResetCodeVerificationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ResetCodeVerificationRequestMultiError, or nil if none found.
func (m *ResetCodeVerificationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ResetCodeVerificationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ResetCode

	if len(errors) > 0 {
		return ResetCodeVerificationRequestMultiError(errors)
	}

	return nil
}

// ResetCodeVerificationRequestMultiError is an error wrapping multiple
// validation errors returned by ResetCodeVerificationRequest.ValidateAll() if
// the designated constraints aren't met.
type ResetCodeVerificationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ResetCodeVerificationRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ResetCodeVerificationRequestMultiError) AllErrors() []error { return m }

// ResetCodeVerificationRequestValidationError is the validation error returned
// by ResetCodeVerificationRequest.Validate if the designated constraints
// aren't met.
type ResetCodeVerificationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ResetCodeVerificationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ResetCodeVerificationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ResetCodeVerificationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ResetCodeVerificationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ResetCodeVerificationRequestValidationError) ErrorName() string {
	return "ResetCodeVerificationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ResetCodeVerificationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sResetCodeVerificationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ResetCodeVerificationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ResetCodeVerificationRequestValidationError{}

// Validate checks the field values on ResetCodeVerificationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ResetCodeVerificationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ResetCodeVerificationResponse with
// the rules defined in the proto definition for this message. If any rules
// are violated, the result is a list of violation errors wrapped in
// ResetCodeVerificationResponseMultiError, or nil if none found.
func (m *ResetCodeVerificationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ResetCodeVerificationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Email

	if len(errors) > 0 {
		return ResetCodeVerificationResponseMultiError(errors)
	}

	return nil
}

// ResetCodeVerificationResponseMultiError is an error wrapping multiple
// validation errors returned by ResetCodeVerificationResponse.ValidateAll()
// if the designated constraints aren't met.
type ResetCodeVerificationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ResetCodeVerificationResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ResetCodeVerificationResponseMultiError) AllErrors() []error { return m }

// ResetCodeVerificationResponseValidationError is the validation error
// returned by ResetCodeVerificationResponse.Validate if the designated
// constraints aren't met.
type ResetCodeVerificationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ResetCodeVerificationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ResetCodeVerificationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ResetCodeVerificationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ResetCodeVerificationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ResetCodeVerificationResponseValidationError) ErrorName() string {
	return "ResetCodeVerificationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ResetCodeVerificationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sResetCodeVerificationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ResetCodeVerificationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ResetCodeVerificationResponseValidationError{}

// Validate checks the field values on GenerateResetCodeRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GenerateResetCodeRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GenerateResetCodeRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GenerateResetCodeRequestMultiError, or nil if none found.
func (m *GenerateResetCodeRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GenerateResetCodeRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Email

	if len(errors) > 0 {
		return GenerateResetCodeRequestMultiError(errors)
	}

	return nil
}

// GenerateResetCodeRequestMultiError is an error wrapping multiple validation
// errors returned by GenerateResetCodeRequest.ValidateAll() if the designated
// constraints aren't met.
type GenerateResetCodeRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GenerateResetCodeRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GenerateResetCodeRequestMultiError) AllErrors() []error { return m }

// GenerateResetCodeRequestValidationError is the validation error returned by
// GenerateResetCodeRequest.Validate if the designated constraints aren't met.
type GenerateResetCodeRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GenerateResetCodeRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GenerateResetCodeRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GenerateResetCodeRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GenerateResetCodeRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GenerateResetCodeRequestValidationError) ErrorName() string {
	return "GenerateResetCodeRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GenerateResetCodeRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGenerateResetCodeRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GenerateResetCodeRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GenerateResetCodeRequestValidationError{}

// Validate checks the field values on ResetPasswordRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ResetPasswordRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ResetPasswordRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ResetPasswordRequestMultiError, or nil if none found.
func (m *ResetPasswordRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ResetPasswordRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Email

	// no validation rules for ResetCode

	// no validation rules for NewPassword

	if len(errors) > 0 {
		return ResetPasswordRequestMultiError(errors)
	}

	return nil
}

// ResetPasswordRequestMultiError is an error wrapping multiple validation
// errors returned by ResetPasswordRequest.ValidateAll() if the designated
// constraints aren't met.
type ResetPasswordRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ResetPasswordRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ResetPasswordRequestMultiError) AllErrors() []error { return m }

// ResetPasswordRequestValidationError is the validation error returned by
// ResetPasswordRequest.Validate if the designated constraints aren't met.
type ResetPasswordRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ResetPasswordRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ResetPasswordRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ResetPasswordRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ResetPasswordRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ResetPasswordRequestValidationError) ErrorName() string {
	return "ResetPasswordRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ResetPasswordRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sResetPasswordRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ResetPasswordRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ResetPasswordRequestValidationError{}

// Validate checks the field values on Metadata with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Metadata) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Metadata with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in MetadataMultiError, or nil
// if none found.
func (m *Metadata) ValidateAll() error {
	return m.validate(true)
}

func (m *Metadata) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.Fullname != nil {
		// no validation rules for Fullname
	}

	if m.Company != nil {
		// no validation rules for Company
	}

	if m.Country != nil {
		// no validation rules for Country
	}

	if m.JobTitle != nil {
		// no validation rules for JobTitle
	}

	if m.AvatarPath != nil {
		// no validation rules for AvatarPath
	}

	if m.Gender != nil {
		// no validation rules for Gender
	}

	if m.Birthday != nil {
		// no validation rules for Birthday
	}

	if m.Summary != nil {
		// no validation rules for Summary
	}

	if m.Website != nil {
		// no validation rules for Website
	}

	if m.LinkedIn != nil {
		// no validation rules for LinkedIn
	}

	if m.Education != nil {
		// no validation rules for Education
	}

	if len(errors) > 0 {
		return MetadataMultiError(errors)
	}

	return nil
}

// MetadataMultiError is an error wrapping multiple validation errors returned
// by Metadata.ValidateAll() if the designated constraints aren't met.
type MetadataMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m MetadataMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m MetadataMultiError) AllErrors() []error { return m }

// MetadataValidationError is the validation error returned by
// Metadata.Validate if the designated constraints aren't met.
type MetadataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e MetadataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e MetadataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e MetadataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e MetadataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e MetadataValidationError) ErrorName() string { return "MetadataValidationError" }

// Error satisfies the builtin error interface
func (e MetadataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sMetadata.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = MetadataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = MetadataValidationError{}

// Validate checks the field values on ChangePasswordRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ChangePasswordRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ChangePasswordRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ChangePasswordRequestMultiError, or nil if none found.
func (m *ChangePasswordRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ChangePasswordRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for OldPassword

	// no validation rules for NewPassword

	// no validation rules for ConfirmNewPassword

	if len(errors) > 0 {
		return ChangePasswordRequestMultiError(errors)
	}

	return nil
}

// ChangePasswordRequestMultiError is an error wrapping multiple validation
// errors returned by ChangePasswordRequest.ValidateAll() if the designated
// constraints aren't met.
type ChangePasswordRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ChangePasswordRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ChangePasswordRequestMultiError) AllErrors() []error { return m }

// ChangePasswordRequestValidationError is the validation error returned by
// ChangePasswordRequest.Validate if the designated constraints aren't met.
type ChangePasswordRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ChangePasswordRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ChangePasswordRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ChangePasswordRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ChangePasswordRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ChangePasswordRequestValidationError) ErrorName() string {
	return "ChangePasswordRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ChangePasswordRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sChangePasswordRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ChangePasswordRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ChangePasswordRequestValidationError{}

// Validate checks the field values on FindUserByNameRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindUserByNameRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindUserByNameRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindUserByNameRequestMultiError, or nil if none found.
func (m *FindUserByNameRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *FindUserByNameRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	if len(errors) > 0 {
		return FindUserByNameRequestMultiError(errors)
	}

	return nil
}

// FindUserByNameRequestMultiError is an error wrapping multiple validation
// errors returned by FindUserByNameRequest.ValidateAll() if the designated
// constraints aren't met.
type FindUserByNameRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindUserByNameRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindUserByNameRequestMultiError) AllErrors() []error { return m }

// FindUserByNameRequestValidationError is the validation error returned by
// FindUserByNameRequest.Validate if the designated constraints aren't met.
type FindUserByNameRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindUserByNameRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindUserByNameRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindUserByNameRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindUserByNameRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindUserByNameRequestValidationError) ErrorName() string {
	return "FindUserByNameRequestValidationError"
}

// Error satisfies the builtin error interface
func (e FindUserByNameRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindUserByNameRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindUserByNameRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindUserByNameRequestValidationError{}

// Validate checks the field values on FindUserByNameResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindUserByNameResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindUserByNameResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindUserByNameResponseMultiError, or nil if none found.
func (m *FindUserByNameResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *FindUserByNameResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return FindUserByNameResponseMultiError(errors)
	}

	return nil
}

// FindUserByNameResponseMultiError is an error wrapping multiple validation
// errors returned by FindUserByNameResponse.ValidateAll() if the designated
// constraints aren't met.
type FindUserByNameResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindUserByNameResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindUserByNameResponseMultiError) AllErrors() []error { return m }

// FindUserByNameResponseValidationError is the validation error returned by
// FindUserByNameResponse.Validate if the designated constraints aren't met.
type FindUserByNameResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindUserByNameResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindUserByNameResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindUserByNameResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindUserByNameResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindUserByNameResponseValidationError) ErrorName() string {
	return "FindUserByNameResponseValidationError"
}

// Error satisfies the builtin error interface
func (e FindUserByNameResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindUserByNameResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindUserByNameResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindUserByNameResponseValidationError{}

// Validate checks the field values on FindUserByMetadataRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindUserByMetadataRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindUserByMetadataRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindUserByMetadataRequestMultiError, or nil if none found.
func (m *FindUserByMetadataRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *FindUserByMetadataRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	switch v := m.Metadata.(type) {
	case *FindUserByMetadataRequest_Name:
		if v == nil {
			err := FindUserByMetadataRequestValidationError{
				field:  "Metadata",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for Name
	case *FindUserByMetadataRequest_Company:
		if v == nil {
			err := FindUserByMetadataRequestValidationError{
				field:  "Metadata",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for Company
	case *FindUserByMetadataRequest_Country:
		if v == nil {
			err := FindUserByMetadataRequestValidationError{
				field:  "Metadata",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for Country
	case *FindUserByMetadataRequest_JobTitle:
		if v == nil {
			err := FindUserByMetadataRequestValidationError{
				field:  "Metadata",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}
		// no validation rules for JobTitle
	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return FindUserByMetadataRequestMultiError(errors)
	}

	return nil
}

// FindUserByMetadataRequestMultiError is an error wrapping multiple validation
// errors returned by FindUserByMetadataRequest.ValidateAll() if the
// designated constraints aren't met.
type FindUserByMetadataRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindUserByMetadataRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindUserByMetadataRequestMultiError) AllErrors() []error { return m }

// FindUserByMetadataRequestValidationError is the validation error returned by
// FindUserByMetadataRequest.Validate if the designated constraints aren't met.
type FindUserByMetadataRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindUserByMetadataRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindUserByMetadataRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindUserByMetadataRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindUserByMetadataRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindUserByMetadataRequestValidationError) ErrorName() string {
	return "FindUserByMetadataRequestValidationError"
}

// Error satisfies the builtin error interface
func (e FindUserByMetadataRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindUserByMetadataRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindUserByMetadataRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindUserByMetadataRequestValidationError{}

// Validate checks the field values on FindUserByMetadataResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *FindUserByMetadataResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on FindUserByMetadataResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// FindUserByMetadataResponseMultiError, or nil if none found.
func (m *FindUserByMetadataResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *FindUserByMetadataResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return FindUserByMetadataResponseMultiError(errors)
	}

	return nil
}

// FindUserByMetadataResponseMultiError is an error wrapping multiple
// validation errors returned by FindUserByMetadataResponse.ValidateAll() if
// the designated constraints aren't met.
type FindUserByMetadataResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m FindUserByMetadataResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m FindUserByMetadataResponseMultiError) AllErrors() []error { return m }

// FindUserByMetadataResponseValidationError is the validation error returned
// by FindUserByMetadataResponse.Validate if the designated constraints aren't met.
type FindUserByMetadataResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e FindUserByMetadataResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e FindUserByMetadataResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e FindUserByMetadataResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e FindUserByMetadataResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e FindUserByMetadataResponseValidationError) ErrorName() string {
	return "FindUserByMetadataResponseValidationError"
}

// Error satisfies the builtin error interface
func (e FindUserByMetadataResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sFindUserByMetadataResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = FindUserByMetadataResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = FindUserByMetadataResponseValidationError{}

// Validate checks the field values on IncreaseBalanceRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *IncreaseBalanceRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on IncreaseBalanceRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// IncreaseBalanceRequestMultiError, or nil if none found.
func (m *IncreaseBalanceRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *IncreaseBalanceRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	// no validation rules for Amount

	if len(errors) > 0 {
		return IncreaseBalanceRequestMultiError(errors)
	}

	return nil
}

// IncreaseBalanceRequestMultiError is an error wrapping multiple validation
// errors returned by IncreaseBalanceRequest.ValidateAll() if the designated
// constraints aren't met.
type IncreaseBalanceRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m IncreaseBalanceRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m IncreaseBalanceRequestMultiError) AllErrors() []error { return m }

// IncreaseBalanceRequestValidationError is the validation error returned by
// IncreaseBalanceRequest.Validate if the designated constraints aren't met.
type IncreaseBalanceRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e IncreaseBalanceRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e IncreaseBalanceRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e IncreaseBalanceRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e IncreaseBalanceRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e IncreaseBalanceRequestValidationError) ErrorName() string {
	return "IncreaseBalanceRequestValidationError"
}

// Error satisfies the builtin error interface
func (e IncreaseBalanceRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sIncreaseBalanceRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = IncreaseBalanceRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = IncreaseBalanceRequestValidationError{}

// Validate checks the field values on IncreaseBalanceResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *IncreaseBalanceResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on IncreaseBalanceResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// IncreaseBalanceResponseMultiError, or nil if none found.
func (m *IncreaseBalanceResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *IncreaseBalanceResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for NewBalance

	if len(errors) > 0 {
		return IncreaseBalanceResponseMultiError(errors)
	}

	return nil
}

// IncreaseBalanceResponseMultiError is an error wrapping multiple validation
// errors returned by IncreaseBalanceResponse.ValidateAll() if the designated
// constraints aren't met.
type IncreaseBalanceResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m IncreaseBalanceResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m IncreaseBalanceResponseMultiError) AllErrors() []error { return m }

// IncreaseBalanceResponseValidationError is the validation error returned by
// IncreaseBalanceResponse.Validate if the designated constraints aren't met.
type IncreaseBalanceResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e IncreaseBalanceResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e IncreaseBalanceResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e IncreaseBalanceResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e IncreaseBalanceResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e IncreaseBalanceResponseValidationError) ErrorName() string {
	return "IncreaseBalanceResponseValidationError"
}

// Error satisfies the builtin error interface
func (e IncreaseBalanceResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sIncreaseBalanceResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = IncreaseBalanceResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = IncreaseBalanceResponseValidationError{}

// Validate checks the field values on GetBalanceResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetBalanceResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetBalanceResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetBalanceResponseMultiError, or nil if none found.
func (m *GetBalanceResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetBalanceResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Balance

	// no validation rules for IsPremium

	// no validation rules for PremiumExpires

	if len(errors) > 0 {
		return GetBalanceResponseMultiError(errors)
	}

	return nil
}

// GetBalanceResponseMultiError is an error wrapping multiple validation errors
// returned by GetBalanceResponse.ValidateAll() if the designated constraints
// aren't met.
type GetBalanceResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetBalanceResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetBalanceResponseMultiError) AllErrors() []error { return m }

// GetBalanceResponseValidationError is the validation error returned by
// GetBalanceResponse.Validate if the designated constraints aren't met.
type GetBalanceResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetBalanceResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetBalanceResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetBalanceResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetBalanceResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetBalanceResponseValidationError) ErrorName() string {
	return "GetBalanceResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetBalanceResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetBalanceResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetBalanceResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetBalanceResponseValidationError{}

// Validate checks the field values on SetPremiumRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *SetPremiumRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SetPremiumRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SetPremiumRequestMultiError, or nil if none found.
func (m *SetPremiumRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *SetPremiumRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Plan

	if len(errors) > 0 {
		return SetPremiumRequestMultiError(errors)
	}

	return nil
}

// SetPremiumRequestMultiError is an error wrapping multiple validation errors
// returned by SetPremiumRequest.ValidateAll() if the designated constraints
// aren't met.
type SetPremiumRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SetPremiumRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SetPremiumRequestMultiError) AllErrors() []error { return m }

// SetPremiumRequestValidationError is the validation error returned by
// SetPremiumRequest.Validate if the designated constraints aren't met.
type SetPremiumRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SetPremiumRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SetPremiumRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SetPremiumRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SetPremiumRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SetPremiumRequestValidationError) ErrorName() string {
	return "SetPremiumRequestValidationError"
}

// Error satisfies the builtin error interface
func (e SetPremiumRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSetPremiumRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SetPremiumRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SetPremiumRequestValidationError{}

// Validate checks the field values on SetPremiumResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SetPremiumResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SetPremiumResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SetPremiumResponseMultiError, or nil if none found.
func (m *SetPremiumResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *SetPremiumResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Success

	if len(errors) > 0 {
		return SetPremiumResponseMultiError(errors)
	}

	return nil
}

// SetPremiumResponseMultiError is an error wrapping multiple validation errors
// returned by SetPremiumResponse.ValidateAll() if the designated constraints
// aren't met.
type SetPremiumResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SetPremiumResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SetPremiumResponseMultiError) AllErrors() []error { return m }

// SetPremiumResponseValidationError is the validation error returned by
// SetPremiumResponse.Validate if the designated constraints aren't met.
type SetPremiumResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SetPremiumResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SetPremiumResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SetPremiumResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SetPremiumResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SetPremiumResponseValidationError) ErrorName() string {
	return "SetPremiumResponseValidationError"
}

// Error satisfies the builtin error interface
func (e SetPremiumResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSetPremiumResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SetPremiumResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SetPremiumResponseValidationError{}

// Validate checks the field values on GetBalanceRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetBalanceRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetBalanceRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetBalanceRequestMultiError, or nil if none found.
func (m *GetBalanceRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetBalanceRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	if len(errors) > 0 {
		return GetBalanceRequestMultiError(errors)
	}

	return nil
}

// GetBalanceRequestMultiError is an error wrapping multiple validation errors
// returned by GetBalanceRequest.ValidateAll() if the designated constraints
// aren't met.
type GetBalanceRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetBalanceRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetBalanceRequestMultiError) AllErrors() []error { return m }

// GetBalanceRequestValidationError is the validation error returned by
// GetBalanceRequest.Validate if the designated constraints aren't met.
type GetBalanceRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetBalanceRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetBalanceRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetBalanceRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetBalanceRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetBalanceRequestValidationError) ErrorName() string {
	return "GetBalanceRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetBalanceRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetBalanceRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetBalanceRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetBalanceRequestValidationError{}

// Validate checks the field values on DecreaseBalanceRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DecreaseBalanceRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DecreaseBalanceRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DecreaseBalanceRequestMultiError, or nil if none found.
func (m *DecreaseBalanceRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DecreaseBalanceRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for UserId

	// no validation rules for Amount

	if len(errors) > 0 {
		return DecreaseBalanceRequestMultiError(errors)
	}

	return nil
}

// DecreaseBalanceRequestMultiError is an error wrapping multiple validation
// errors returned by DecreaseBalanceRequest.ValidateAll() if the designated
// constraints aren't met.
type DecreaseBalanceRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DecreaseBalanceRequestMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DecreaseBalanceRequestMultiError) AllErrors() []error { return m }

// DecreaseBalanceRequestValidationError is the validation error returned by
// DecreaseBalanceRequest.Validate if the designated constraints aren't met.
type DecreaseBalanceRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DecreaseBalanceRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DecreaseBalanceRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DecreaseBalanceRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DecreaseBalanceRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DecreaseBalanceRequestValidationError) ErrorName() string {
	return "DecreaseBalanceRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DecreaseBalanceRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDecreaseBalanceRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DecreaseBalanceRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DecreaseBalanceRequestValidationError{}

// Validate checks the field values on DecreaseBalanceResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DecreaseBalanceResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DecreaseBalanceResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DecreaseBalanceResponseMultiError, or nil if none found.
func (m *DecreaseBalanceResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DecreaseBalanceResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Success

	if len(errors) > 0 {
		return DecreaseBalanceResponseMultiError(errors)
	}

	return nil
}

// DecreaseBalanceResponseMultiError is an error wrapping multiple validation
// errors returned by DecreaseBalanceResponse.ValidateAll() if the designated
// constraints aren't met.
type DecreaseBalanceResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DecreaseBalanceResponseMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DecreaseBalanceResponseMultiError) AllErrors() []error { return m }

// DecreaseBalanceResponseValidationError is the validation error returned by
// DecreaseBalanceResponse.Validate if the designated constraints aren't met.
type DecreaseBalanceResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DecreaseBalanceResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DecreaseBalanceResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DecreaseBalanceResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DecreaseBalanceResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DecreaseBalanceResponseValidationError) ErrorName() string {
	return "DecreaseBalanceResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DecreaseBalanceResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDecreaseBalanceResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DecreaseBalanceResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DecreaseBalanceResponseValidationError{}

// Validate checks the field values on SignUpRequest_Local with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SignUpRequest_Local) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignUpRequest_Local with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SignUpRequest_LocalMultiError, or nil if none found.
func (m *SignUpRequest_Local) ValidateAll() error {
	return m.validate(true)
}

func (m *SignUpRequest_Local) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Username

	// no validation rules for Password

	// no validation rules for ConfirmPassword

	// no validation rules for Email

	// no validation rules for Otp

	if len(errors) > 0 {
		return SignUpRequest_LocalMultiError(errors)
	}

	return nil
}

// SignUpRequest_LocalMultiError is an error wrapping multiple validation
// errors returned by SignUpRequest_Local.ValidateAll() if the designated
// constraints aren't met.
type SignUpRequest_LocalMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignUpRequest_LocalMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignUpRequest_LocalMultiError) AllErrors() []error { return m }

// SignUpRequest_LocalValidationError is the validation error returned by
// SignUpRequest_Local.Validate if the designated constraints aren't met.
type SignUpRequest_LocalValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignUpRequest_LocalValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignUpRequest_LocalValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignUpRequest_LocalValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignUpRequest_LocalValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignUpRequest_LocalValidationError) ErrorName() string {
	return "SignUpRequest_LocalValidationError"
}

// Error satisfies the builtin error interface
func (e SignUpRequest_LocalValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignUpRequest_Local.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignUpRequest_LocalValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignUpRequest_LocalValidationError{}

// Validate checks the field values on SignUpRequest_Google with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SignUpRequest_Google) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignUpRequest_Google with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SignUpRequest_GoogleMultiError, or nil if none found.
func (m *SignUpRequest_Google) ValidateAll() error {
	return m.validate(true)
}

func (m *SignUpRequest_Google) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Credential

	if len(errors) > 0 {
		return SignUpRequest_GoogleMultiError(errors)
	}

	return nil
}

// SignUpRequest_GoogleMultiError is an error wrapping multiple validation
// errors returned by SignUpRequest_Google.ValidateAll() if the designated
// constraints aren't met.
type SignUpRequest_GoogleMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignUpRequest_GoogleMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignUpRequest_GoogleMultiError) AllErrors() []error { return m }

// SignUpRequest_GoogleValidationError is the validation error returned by
// SignUpRequest_Google.Validate if the designated constraints aren't met.
type SignUpRequest_GoogleValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignUpRequest_GoogleValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignUpRequest_GoogleValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignUpRequest_GoogleValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignUpRequest_GoogleValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignUpRequest_GoogleValidationError) ErrorName() string {
	return "SignUpRequest_GoogleValidationError"
}

// Error satisfies the builtin error interface
func (e SignUpRequest_GoogleValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignUpRequest_Google.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignUpRequest_GoogleValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignUpRequest_GoogleValidationError{}

// Validate checks the field values on SignInRequest_Local with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SignInRequest_Local) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignInRequest_Local with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SignInRequest_LocalMultiError, or nil if none found.
func (m *SignInRequest_Local) ValidateAll() error {
	return m.validate(true)
}

func (m *SignInRequest_Local) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Email

	// no validation rules for Password

	if len(errors) > 0 {
		return SignInRequest_LocalMultiError(errors)
	}

	return nil
}

// SignInRequest_LocalMultiError is an error wrapping multiple validation
// errors returned by SignInRequest_Local.ValidateAll() if the designated
// constraints aren't met.
type SignInRequest_LocalMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignInRequest_LocalMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignInRequest_LocalMultiError) AllErrors() []error { return m }

// SignInRequest_LocalValidationError is the validation error returned by
// SignInRequest_Local.Validate if the designated constraints aren't met.
type SignInRequest_LocalValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignInRequest_LocalValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignInRequest_LocalValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignInRequest_LocalValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignInRequest_LocalValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignInRequest_LocalValidationError) ErrorName() string {
	return "SignInRequest_LocalValidationError"
}

// Error satisfies the builtin error interface
func (e SignInRequest_LocalValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignInRequest_Local.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignInRequest_LocalValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignInRequest_LocalValidationError{}

// Validate checks the field values on SignInRequest_Google with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *SignInRequest_Google) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on SignInRequest_Google with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// SignInRequest_GoogleMultiError, or nil if none found.
func (m *SignInRequest_Google) ValidateAll() error {
	return m.validate(true)
}

func (m *SignInRequest_Google) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Credential

	if len(errors) > 0 {
		return SignInRequest_GoogleMultiError(errors)
	}

	return nil
}

// SignInRequest_GoogleMultiError is an error wrapping multiple validation
// errors returned by SignInRequest_Google.ValidateAll() if the designated
// constraints aren't met.
type SignInRequest_GoogleMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m SignInRequest_GoogleMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m SignInRequest_GoogleMultiError) AllErrors() []error { return m }

// SignInRequest_GoogleValidationError is the validation error returned by
// SignInRequest_Google.Validate if the designated constraints aren't met.
type SignInRequest_GoogleValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e SignInRequest_GoogleValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e SignInRequest_GoogleValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e SignInRequest_GoogleValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e SignInRequest_GoogleValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e SignInRequest_GoogleValidationError) ErrorName() string {
	return "SignInRequest_GoogleValidationError"
}

// Error satisfies the builtin error interface
func (e SignInRequest_GoogleValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sSignInRequest_Google.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = SignInRequest_GoogleValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = SignInRequest_GoogleValidationError{}
